packages/api/amplify_api/lib/src/graphql/ws/web_socket_connection.dart

```diff
@@ -74,15 +94,55 @@ class WebSocketConnection implements Closeable {
       Stream<dynamic> stream) {
     return stream.transform(const WebSocketMessageStreamTransformer()).listen(
       _onData,
-      cancelOnError: true,
+      onDone: () {
+        _logger.verbose('Web socket connection done.');
+
+        _resetConnectionInit();
+        _resetConnectionVariables();
+      },
       onError: (Object e) {
+        if (!_connectionReady.isCompleted) _connectionReady.completeError(e);
         _connectionError(
-          ApiException('Connection failed.', underlyingException: e.toString()),
+          ApiException('Connection failed.', underlyingException: e),
         );
+        _hubEventsController?.add(SubscriptionHubEvent.failed());
+
+        if (_hasNetwork &&
+            _hasConnectionBroken &&
```

Comment:
Not sure we want to check _hasConnectionBroken here since that will only be true if _hasNetwork is false really.

Actually, in what situation would we get here with network and not already be retrying?

Reply:
I see what you're saying. _hasConnectionBroken indicates that a connection existed at one point, but was interrupted & a retry attempt hasn't been made yet. If it's false, then we've already tried to recover. 

That being said, since the called out code lives in the onError handler I agree _hasConnectionBroken doesn't need to be checked. Network related interruptions are being handled else where. The goal here is to recover from errors. 

Good catch.